### DISCUSSION TITLE
How to obtain autonum key value while in before_flush

### DISCUSSION THREAD

**Discussion by valhuber** [sqlalchemy/sqlalchemy#10667] [Usage Questions] [answered]
I use `before_flush` for business logic (derivations, constraints), and it works very well.

A new requirement is to prepare a message (or API) to send an inserted row to a remote system.  The inserted row has an autonum key, so its value is None.  Not unexpected, but the remote system needs to know the row id.

I am seeking a way to flush the cache, but not raise `before_flush` since I'm already in it.

I tried `session.flush()`, but, as I feared/expected, got `sqlalchemy.exc.InvalidRequestError: Session is already flushing`.  That is certainly true... is there a "light flush" so I can obtain the key for the remote system?

**Comment by zzzeek** [comment]
if you dont run the INSERT statement directly or otherwise consult a sequence in that event, you wont have that rownum until at the earliest the after_insert event.

you can use that event, or an event like after_flush(). all of these still take place before any transaction is committed so the data is not yet durable / available from the outside.

**Comment by valhuber** [comment]
Awesome, thanks for the info!  Sounds perfect, will give it a shot

Many thanks,
Val

**Reply by CaselIT** [reply, marked as answer]
You can also use the `session.info` dict to pass state from one event to the other

**Reply by valhuber** [reply]
This has worked well (meant to get back earlier).

I thought the [doc was slightly misleading](https://docs.sqlalchemy.org/en/20/orm/session_events.html#after-flush)... it did not suggest that after_flush sees autonum fields.  You might want to add a note for that.
